### PR TITLE
[HOMEWORK-TASK-3_2] Refactor react-router logic

### PR DESCRIPTION
**Pull request by vadymshukr**
WHAT
All logic connected to react-router collected in one file

HOW
Created custom function for site links navigation
All native methods of react-router removed from application
All imports from react-router collected in one file

WHY
Refactored react routing for more easy maintanance of application in case if react-router library will be change to something else

### CODE REVIEW

**Review by maxzag — APPROVED** [review]

File: src/helpers/funcs.ts [hunk 1]
```diff
@@ -1,3 +1,10 @@
 export const makeStringWithError = (value: string, length: number) => {
     return value.length <= length ? value : `${value.slice(0, length)}<small>${value.slice(length)}</small>`
-}
\ No newline at end of file
+}
+
+
+export const changeRoute = (route: string) => {
+    history.pushState({}, '', "/" + route)
```
Comment (Rastopyr) [inline src/helpers/funcs.ts]: why make push state twice?

**Comment by Rastopyr** [comment]
Here is some problems with dependencies in app, let's make review of your PR on the next call?

File: src/pages/singleNote/single-note.tsx [hunk 2]
```diff
@@ -16,10 +16,8 @@ const linkEventListener = (e: MouseEvent, link: HTMLAnchorElement) =>
 }
 
 export function SingleNote() {
-  const route = useParams();
-  const {id} = route;
+  const id = window.location.pathname.split('/')[1];
```
Comment (Rastopyr) [inline src/pages/singleNote/single-note.tsx]: better to use some  state from providers

**Review by Rastopyr — CHANGES_REQUESTED** [review]

File: src/helpers/funcs.ts (hunk 1 above)
Comment (vadymshukr) [reply src/helpers/funcs.ts]: It's some kind of hack for react router. First call - we change url but is no redirect to page, second call - we push same adres to history, and then go back one step before for finally page render.
All other ways (like location.href) makes pages reload.

Comment (Rastopyr) [reply src/helpers/funcs.ts]: you can provide your own wrapper for react router, and encapsulate navigation through it

Comment (Rastopyr) [reply src/helpers/funcs.ts]: this is more preferable way to do  that, instead of hacks with history API

Comment (Rastopyr) [reply src/helpers/funcs.ts]: hack is not testable as usual, and very hard to extend it behavior

File: src/contexts/notes-list-context.tsx [hunk 3]
```diff
@@ -61,9 +61,9 @@ export function ReduxWrapper({children} : PropsWithChildren<any>){
 export function useNoteListState() {
     const [state, dispatch] = useContext(NotesListContext)
     const [actions] = useState(() => ({
-        setInitialNotes: (payload: any) => dispatch(setInitialNotes(payload)),
-        addNewNote: (payload: any) => dispatch(addNewNote(payload)),
-        editNote: (payload: any) => dispatch(editNote(payload)),
+        setInitialNotes: (payload: NotesType[]) => dispatch(setInitialNotes(payload)),
+        addNewNote: (payload: NotesType) => dispatch(addNewNote(payload)),
+        editNote: (payload: NotesType) => dispatch(editNote(payload)),
         deleteNote:(payload: any) => dispatch(deleteNote(payload))
```
Comment (maxzag) [inline src/contexts/notes-list-context.tsx]: remove any

**Review by Rastopyr — APPROVED** [review]

**Pull request merged**